File: app/src/main/java/com/gelostech/pocketbartender/fragments/SearchFragment.java
package com.gelostech.pocketbartender.fragments;


import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Parcel;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.arlib.floatingsearchview.FloatingSearchView;
import com.arlib.floatingsearchview.suggestions.SearchSuggestionsAdapter;
import com.arlib.floatingsearchview.suggestions.model.SearchSuggestion;
import com.gelostech.pocketbartender.R;
import com.gelostech.pocketbartender.activities.CocktailActivity;
import com.gelostech.pocketbartender.adapters.SearchAdapter;
import com.gelostech.pocketbartender.commoners.BartenderSingleton;
import com.gelostech.pocketbartender.commoners.BartenderUtil;
import com.gelostech.pocketbartender.commoners.PillNode;
import com.gelostech.pocketbartender.models.HomeModel;
import com.mikepenz.iconics.IconicsDrawable;
import com.mikepenz.ionicons_typeface_library.Ionicons;

import org.jetbrains.annotations.NotNull;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * A simple {@link Fragment} subclass.
 */
public class SearchFragment extends Fragment implements FloatingSearchView.OnQueryChangeListener, FloatingSearchView.OnSearchListener,
        FloatingSearchView.OnHomeActionClickListener, FloatingSearchView.OnMenuItemClickListener, SearchSuggestionsAdapter.OnBindSuggestionCallback{
    private View view;
    private SearchAdapter searchAdapter;

    @BindView(R.id.search_view) FloatingSearchView searchView;
    @BindView(R.id.search_rv) RecyclerView rv;
    @BindView(R.id.search_pill) PillNode pillNode;

    public SearchFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        if(view == null){
            view = inflater.inflate(R.layout.fragment_search, container, false);
            ButterKnife.bind(this, view);

            initViews();
        }

        return view;
    }

    private void initViews(){
        searchView.setClearBtnColor(ContextCompat.getColor(getActivity(), R.color.colorAccent));
        searchView.setOnQueryChangeListener(this);
        searchView.setOnMenuItemClickListener(this);
        searchView.setOnBindSuggestionCallback(this);
        searchView.setOnHomeActionClickListener(this);
        searchView.setOnSearchListener(this);

        rv.setHasFixedSize(true);
        rv.setLayoutManager(new LinearLayoutManager(getActivity()));
        rv.addItemDecoration(new SimpleDividerItemDecoration(getActivity()));

        searchAdapter = new SearchAdapter(getActivity(), new SearchAdapter.OnSearchResultClick() {
            @Override
            public void onSearchResultClick(HomeModel model) {
                Intent intent = new Intent(getActivity(), CocktailActivity.class);
                intent.putExtra("cocktail", model);
                startActivity(intent);
                getActivity().overridePendingTransition(R.anim.enter_b, R.anim.exit_a);
            }
        });
        rv.setAdapter(searchAdapter);

        pillNode.addPills("name", "ingredient", "ordinary drink", "cocktail");
        pillNode.setPillListener(new PillNode.PillListener() {
            @Override
            public void onPillClicked(@NotNull String pill, int pos) {

            }
        });

    }


    private void searchCocktails(String name){
        JsonObjectRequest fetchCocktails = new JsonObjectRequest(Request.Method.GET, BartenderUtil.SEARCH_URL + name, null, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                try {
                    JSONArray cocktailArray = response.getJSONArray("drinks");

                    for(int i=0; i<cocktailArray.length(); i++){
                        JSONObject cocktailObject = cocktailArray.getJSONObject(i);

                        HomeModel cocktail = new HomeModel();
                        cocktail.setName(cocktailObject.getString("strDrink"));
                        cocktail.setImageUrl(cocktailObject.getString("strDrinkThumb"));
                        cocktail.setId(cocktailObject.getInt("idDrink"));

                        searchAdapter.addCocktail(cocktail);
                    }


                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {

            }
        });


        BartenderSingleton.getInstance(getActivity()).addToRequestQueue(fetchCocktails);
    }


    @Override
    public void onSearchTextChanged(String oldQuery, String newQuery) {
        //searchView.swapSuggestions(suggestions);
    }

    @Override
    public void onHomeClicked() {
        searchView.clearQuery();
        Toast.makeText(getActivity(), "Back clicked", Toast.LENGTH_SHORT).show();
    }


    @Override
    public void onActionMenuItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.filter_search:
                //item.setIcon(new IconicsDrawable(getActivity()).icon(Ionicons.Icon.ion_ios_settings).sizeDp(20).color(ContextCompat.getColor(getActivity(),R.color.colorAccent)));
                //Toast.makeText(getActivity(), "sbdfbdn", Toast.LENGTH_SHORT).show();
                break;

            default:
                break;
        }
    }

    @Override
    public void onBindSuggestion(View suggestionView, ImageView leftIcon, TextView textView, SearchSuggestion item, int itemPosition) {

    }


    @Override
    public void onSuggestionClicked(SearchSuggestion searchSuggestion) {

    }

    @Override
    public void onSearchAction(String currentQuery) {
        searchAdapter.clear();
        searchCocktails(currentQuery);
    }

    public class SimpleDividerItemDecoration extends RecyclerView.ItemDecoration {
        private Drawable mDivider;

        public SimpleDividerItemDecoration(Context context) {
            mDivider = ContextCompat.getDrawable(context,R.drawable.recycler_divider);
        }

        @Override
        public void onDrawOver(Canvas c, RecyclerView parent, RecyclerView.State state) {
            int left = parent.getPaddingLeft();
            int right = parent.getWidth() - parent.getPaddingRight();

            int childCount = parent.getChildCount();
            for (int i = 0; i < childCount; i++) {
                View child = parent.getChildAt(i);

                RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();

                int top = child.getBottom() + params.bottomMargin;
                int bottom = top + mDivider.getIntrinsicHeight();

                mDivider.setBounds(left, top, right, bottom);
                mDivider.draw(c);
            }
        }
    }

}

File: app/src/main/java/com/gelostech/pocketbartender/fragments/SearchFragment.kt
package com.gelostech.pocketbartender.fragments


import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.Toolbar
import android.view.*
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.gelostech.pocketbartender.R
import com.gelostech.pocketbartender.activities.CocktailActivity
import com.gelostech.pocketbartender.activities.MainActivity
import com.gelostech.pocketbartender.adapters.SearchAdapter
import com.gelostech.pocketbartender.commoners.BartenderSingleton
import com.gelostech.pocketbartender.commoners.BartenderUtil
import com.gelostech.pocketbartender.models.HomeModel
import com.gelostech.pocketbartender.utils.RecyclerFormatter
import com.gelostech.pocketbartender.utils.setFont
import com.miguelcatalan.materialsearchview.MaterialSearchView
import com.mikepenz.iconics.IconicsDrawable
import com.mikepenz.ionicons_typeface_library.Ionicons
import kotlinx.android.synthetic.main.fragment_search.*
import kotlinx.android.synthetic.main.fragment_search.view.*
import org.json.JSONException
import android.os.Handler
import android.util.Log
import android.view.inputmethod.InputMethodManager
import com.gelostech.pocketbartender.commoners.DatabaseHelper
import com.gelostech.pocketbartender.utils.snack
import org.jetbrains.anko.toast


/**
 * A simple [Fragment] subclass.
 */
class SearchFragment : Fragment(), SearchAdapter.OnSearchResultClick, SearchAdapter.OnSearchHistoryClick, MaterialSearchView.OnQueryTextListener {

    private var searchAdapter: SearchAdapter? = null
    private var query: String? = null
    private var myHandler = Handler()
    private val WAIT = 1300L
    private lateinit var db: DatabaseHelper

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_search, container, false)
        db = DatabaseHelper(context)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews(view)
        searchHistory()
    }

    private fun initViews(v: View) {
        (activity as MainActivity).setSupportActionBar(v.searchToolbar)
        (activity as MainActivity).supportActionBar?.title = null
        v.searchToolbarTitle.setFont(BartenderUtil.FONT)
        v.searchToolbarTitle.text = getString(R.string.search)
        setupToolbar(v.searchToolbar, v)

        v.searchRv.setHasFixedSize(true)
        v.searchRv.layoutManager = LinearLayoutManager(context)
        v.searchRv.addItemDecoration(RecyclerFormatter.SimpleDividerItemDecoration(activity!!))

        searchAdapter = SearchAdapter(activity, this, this)
        v.searchRv.adapter = searchAdapter

        v.searchingIcon.setImageDrawable(IconicsDrawable(activity!!).icon(Ionicons.Icon.ion_ios_search).sizeDp(22).color(ContextCompat.getColor(activity!!, R.color.dark_gray)))
    }

    private fun setupToolbar(toolbar: Toolbar, v: View) {
        toolbar.inflateMenu(R.menu.search_toolbar_menu)
        val item = toolbar.menu.findItem(R.id.action_search)
        item?.icon = IconicsDrawable(activity!!).icon(Ionicons.Icon.ion_ios_search_strong).sizeDp(22).color(ContextCompat.getColor(activity!!, R.color.colorAccent))
        v.searchView.setMenuItem(item)
        v.searchView.setOnQueryTextListener(this)

    }

    private fun searchHistory(){
        val history = db.fetchHistory()

        if(history.size > 0){
            with(history) {
                forEach { searchAdapter!!.addCocktail(it) }
            }
        } else {
            view?.noSearchHistory!!.visibility = View.VISIBLE
        }
    }

    private fun searchCocktails(name: String) {
        searchAdapter!!.clear()

        val fetchCocktails = JsonObjectRequest(Request.Method.GET, BartenderUtil.SEARCH_URL + name, null, Response.Listener { response ->
            try {
                val cocktailArray = response.getJSONArray("drinks")

                view?.noSearchHistory!!.visibility = View.GONE

                for (i in 0 until cocktailArray.length()) {
                    val cocktailObject = cocktailArray.getJSONObject(i)

                    val cocktail = HomeModel()
                    cocktail.name = cocktailObject.getString("strDrink")
                    cocktail.imageUrl = cocktailObject.getString("strDrinkThumb")
                    cocktail.id = cocktailObject.getInt("idDrink")
                    cocktail.type = 1

                    searchAdapter!!.addCocktail(cocktail)
                    Handler().postDelayed({view?.searchingLayout!!.visibility = View.GONE}, 1500)
                }

            } catch (e: JSONException) {
                Log.d(javaClass.simpleName, "Json Exception: $e.localizedMessage")
                Handler().postDelayed({view?.searchingLayout!!.visibility = View.GONE}, 1500)
            }
        }, Response.ErrorListener {
            Log.d(javaClass.simpleName, "Response error: $it.localizedMessage")
            Handler().postDelayed({view?.searchingLayout!!.visibility = View.GONE}, 1500)
        })

        BartenderSingleton.getInstance(activity).addToRequestQueue(fetchCocktails)
    }

    override fun onSearchResultClick(model: HomeModel) {
        db.addHistory(model.id.toString(), model.name, model.imageUrl)

        val intent = Intent(activity, CocktailActivity::class.java)
        intent.putExtra("cocktail", model)
        startActivity(intent)
        activity!!.overridePendingTransition(R.anim.enter_b, R.anim.exit_a)
    }

    override fun onSearchHistoryClick(model: HomeModel?) {
        val intent = Intent(activity, CocktailActivity::class.java)
        intent.putExtra("cocktail", model)
        startActivity(intent)
        activity!!.overridePendingTransition(R.anim.enter_b, R.anim.exit_a)
    }

    override fun onQueryTextSubmit(query: String?): Boolean {
        view?.searchingLayout!!.visibility = View.VISIBLE

        Handler().postDelayed({searchCocktails(query!!.trim())}, 2000)
        val imm = activity!!.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(view?.applicationWindowToken, 0)
        return true
    }

    override fun onQueryTextChange(newText: String?): Boolean {
        newText?.let {
            query = it.trim()
            restart() }
        return true
    }

    fun onBackPressed():Boolean = when(view?.searchView!!.isSearchOpen){
        true->{
            searchView.closeSearch()
            stop()
            false
        }
        else->true
    }

    private var myRunnable: Runnable = Runnable {
        if(query!!.isNotEmpty() && query!!.isNotBlank()){
            view?.searchingLayout!!.visibility = View.VISIBLE
            searchCocktails(query!!)
        }
    }


    fun start() {
        myHandler.postDelayed(myRunnable, WAIT)
    }

    private fun stop() {
        myHandler.removeCallbacks(myRunnable)
    }

    private fun restart() {
        myHandler.removeCallbacks(myRunnable)
        myHandler.postDelayed(myRunnable, WAIT)
    }

    override fun onPause() {
        super.onPause()
        stop()
    }


}
